stingray/pulse/search.py

```diff
@@ -157,33 +157,166 @@ def search_best_peaks(x, stat, threshold):
+        bove threshold.
     """
     stat = np.asarray(stat)
     x = np.asarray(x)
     peaks = stat >= threshold
     regions = contiguous_regions(peaks)
     if len(regions) == 0:
-        return []
+        return [], []
     best_x = np.zeros(len(regions))
+    best_stat = np.zeros(len(regions))
     for i, r in enumerate(regions):
-        best_x[i] = x[r[0]:r[1]][np.argmax(stat[r[0]:r[1]])]
+        stat_filt = stat[r[0]:r[1]]
+        x_filt = x[r[0]:r[1]]
+        max_arg = np.argmax(stat_filt)
+        best_stat[i] = stat_filt[max_arg]
+        best_x[i] = x_filt[max_arg]
+
+    order = np.argsort(best_stat)[::-1]
+
+    return best_x[order], best_stat[order]
+
+
+def phaseogram(times, f, nph=128, nt=32, ph0=0, mjdref=None,
+               fdot=0, fddot=0, pepoch=None, return_plot=False,
+               out_filename='phaseogram.png'):
+    """
+    Calculate and plot the phaseogram of a pulsar observation.
+
+    The phaseogram is a 2-D histogram where the x axis is the pulse phase and
+    the y axis is the time. It shows how the pulse phase changes with time, and
+    it is very useful to see if the pulse solution is correct and/or if there
+    are additional frequency derivatives appearing in the data (due to spin up
+    or down, or even orbital motion)
+
+    Parameters
+    ----------
+    times : array
+        Event arrival times
+    f : float
+        Pulse frequency
+
+    Other parameters
+    ----------------
+    nph : int
+        Number of phase bins
+    nt : int
+        Number of time bins
+    ph0 : float
+        The starting phase of the pulse
+    mjdref : float
+        MJD reference time. If given, the y axis of the plot will be in MJDs,
+        otherwise it will be in seconds.
+    fdot : float
+        First frequency derivative
+    fddot : float
+        Second frequency derivative
+    pepoch : float
+        If the input pulse solution is referred to a given time, give it here.
+        It has no effect (just a phase shift of the pulse) if `fdot` is zero.
+        if `mjdref` is specified, pepoch MUST be in MJD
+    return_plot : bool
+        Return the axes in the additional_info, and don't close the plot, so
+        that the user can add information to it.
+    out_filename : str
+        Output file name for the plot
+
+    Returns
+    -------
+    phaseogr : 2-D matrix
+        The phaseogram
+    phases : array-like
+        The x axis of the phaseogram (the x bins of the histogram),
+        corresponding to the pulse phase in each column
+    times : array-like
+        The y axis of the phaseogram (the y bins of the histogram),
+        corresponding to the time at each row
+    additional_info : dict
+        Additional information, like the pulse profile and the axes to modify
+        the plot (the latter, only if `return_plot` is True)
+    """
+
+    import matplotlib.pyplot as plt
+    import matplotlib as mpl
```

Comment:
Three comments on the plotting:
* You only need matplotlib if `return_plot=True`, right? So why not put this in an if-statement to make sure the function executes when `return_plot=False`, even if matplotlib isn't installed?
* the plotting functionality seems fairly complex. Why not pull it into its own separate function, then we don't have to worry about whether the function executes depending on matplotlib being installed?
* These days, I like a syntax where you can pass in an `ax` object into a function and it will draw the plot inside. Then the user can wrap the plotting function into their own code, which plots to file if necessary. I often want to fiddle with the figures that come out of these kinds of functions before saving them to file.